src/sentry/ratelimits/leaky_bucket.py

```diff
@@ -0,0 +1,203 @@
+
+        return LeakyBucketLimitInfo(
+            self.burst_limit,
+            self.drip_rate,
+            last_drip,
+            current_level,
+            max(0, (current_level - self.burst_limit) / self.drip_rate),
+        )
+
+    def decorator(
+        self,
+        key_override: str | None = None,
+        limited_handler: Callable[[LeakyBucketLimitInfo, dict[str, Any]], Any] | None = None,
+        raise_exception: bool = False,
+    ) -> Callable[[Any], Any]:
+        """
+        Decorator to limit the rate of requests
+
+        key_override: a string that will be used as the key to identify the rate limit,
+                      if not provided fully qualified function name will be used
+
+        limited_handler: a callback function that will be called when the rate limit is exceeded,
+                            it will receive the LeakyBucketLimitInfo instance as an argument,
+                            as well as original args and kwargs in a dictionary
+                        return value will be returned as the result of the decorated function
+
+        raise_exception: if True, LeakyBucketRateLimiter.LimitExceeded will be raised when the rate limit is exceeded
+
+
+        if both limited_callback and raise_exception are provided, limited_callback will take precedence
+        if neither limited_callback nor raise_exception is provided, the decorated function will silently
+        be ignored when the rate limit is exceeded
+
+
+
+        usage:
+
+            - basic usage:
+            ```
+            limiter = LeakyBucketRateLimiter(burst_limit=10, drip_rate=1)
+
+            @limiter()
+            def my_function():
+                do_something()
+            ```
+
+
+            ```
+            limiter = LeakyBucketRateLimiter(burst_limit=10, drip_rate=1)
+
+            @limiter(key)
+            def my_function():
+                do_something()
+            ```
+
+            - raising an exception when limited:
+            ```
+            try:
+                @limiter(key, raise_exception=True)
+                def my_function():
+                    do_something()
+            except LeakyBucketRateLimiter.LimitExceeded as e:
+                print("Rate limited, needs to wait for {e.wait_time} seconds")
+            ```
+
+            - providing a handler function:
+            ```
+            def my_limited_handler(info, context):
+                print(f"Rate limited, needs to wait for {info.wait_time} seconds")
+                print(f"Original args: {context['args']}")
+                print(f"Original kwargs: {context['kwargs']}")
+                return None
+
+            @limiter(key, limited_handler=my_limited_handler)
+            def my_function():
+                rv = do_something()
+            ```
+
+        """
+
+        def decorator(func: Callable[[Any], Any]) -> Callable[[Any], Any]:
+            @functools.wraps(func)
```

Comment:
this should use `Callable[P, R] -> Callable[P, R]` to preserve the signature (`P` being the conventional name for a `ParamSpec` and `R` being a TypeVar for the return value): https://youtu.be/fwZoxWyMGM8

Reply:
@asottile-sentry https://github.com/getsentry/sentry/pull/74687